### DISCUSSION TITLE
@FindAll and @FindBys Similar Thing In WebdriverIO

### DISCUSSION THREAD

**Discussion by anupamtomar** [webdriverio/webdriverio#9866] [Q&A]
How can we achieve Selenium's **@FindAll and @FindBys** or something similar to it in WebdriverIO?

**Comment by christian-bromann** [comment]
Please see the project docs:
- https://webdriver.io/docs/selectors
- https://webdriver.io/docs/api/browser/$$

**Reply by anupamtomar** [reply]
I have gone through the above shared links but Did not find exactly what i am looking for, 

We use these these two annotations in those cases when we have more than a single criteria to to identify one or more WebElement objects.

**Usage:**

**@FindBys :** When the required WebElement objects need to match all of the given criteria use @FindBys annotation.
To locate a web element with more than one search criteria, you can use @FindBys annotation. This annotation locates the web element by using the **AND** condition on the search criteria. In simple words, @FindBys uses multiple @FindBy for each search criteria.

```
@FindBys( {
  @FindBy(className = "class1")
  @FindBy(className = "class2")
} )
List<WebElement> elementsWithBoth_class1ANDclass2;

//Here List elementsWithBoth_class1ANDclass2 will contain all the WebElements which satisfies both criteria.
```

```
@FindBys({
  @FindBy(class="custom-control-check-box"),
  @FindBy(id="game-chk-box")
})

WebElement chkBox;

//Here WebElement chkBox will contain an WebElement which satisfies both criteria.
```

**@FindAll :** When required WebElement objects need to match at least one of the given criteria use @FindAll annotation.
The @FindAll annotation locates the web element using more than one criteria, given that at least one criteria match. Contrary to @FindBys, it uses an **OR** conditional relationship between the multiple @FindBy.

```
@FindAll({
  @FindBy(className = "class1")
  @FindBy(className = "class2")
})
List<WebElement> elementsWithEither_class1ORclass2;

//Here List elementsWithEither_class1ORclass2 will contain all those WebElement that satisfies any one of the criteria.
```

```
@FindAll({
  @FindBy(class="btn-submit", //doesn't match
  @FindBy(class="sbmt-Btn"), //Matches
  @FindBy(class="btn-primary") //doesn't match
})
WebElement submitButton;

//Here it would locate the submit button even though only one criterion matches @FindAll works on one or more criteria.
```


Could you please give an example of both **@FindBys** and **@FindAll** that how can we achieve it exactly in webdriverio, mainly i am looking for example of @FindAll for single WebElement search not a list of webelements (Basically the last example i have given).

**Reply by christian-bromann** [reply]
@anupamtomar sorry but it seems you have a custom way of finding elements which is totally possible to implement in WebdriverIO too with the commands I showed you. I can't give you an example on top of my head how this would need to be implemented. I think you know the best how your locators work and will be able to translate this into WebdriverIO land.

**Reply by anupamtomar** [reply]
Sorry but I haven't customized anything, I have used as it is provided by `org.openqa.selenium`, seems there is no similar, easy, and direct solution for it in WebdriverIO instead of implementing using commands which looks little complex or maybe difficult for new bees or perhaps no similar solution, I don't know. 
anyways... Thanks!

**Reply by christian-bromann** [reply]
> seems there is no similar, easy, and direct solution for it in WebdriverIO instead of implementing using commands which looks little complex or maybe difficult for new bees or perhaps no similar solution, I don't know.

WebdriverIO has a different API so you need to check out the WebdriverIO docs to understand how to achieve the same. There are plenty of learning material out there, e.g. https://www.youtube.com/watch?v=kRkQbeQY8Ho to help you understand the WebdriverIO API.
